modules/gateways/veritrans.php

```diff
@@ -349,21 +349,54 @@ function veritrans_link($params)
         $params['credit_card']['save_card'] = true;
     }
 
-    // Get snap token
     try {
-        // debug
-        // echo "<pre>";
-        // var_dump($params);
-        // echo "</pre>";
-        // exit();
-        $snap_transaction = Veritrans_Snap::createTransaction($params);
-        $snapToken = $snap_transaction->token;
-        $redirect_url = $snap_transaction->redirect_url;
-        // error_log(" ############# TOKEN ::: ".$snapToken);
+        $whmcsApiCommand = 'GetInvoice';
+        $whmcsApiPostParams = array(
+            'invoiceid' => $invoiceId
+        );
+        $invoice = localAPI($whmcsApiCommand, $whmcsApiPostParams);
+        $existingInvoiceNotes = $invoice->notes;
     } catch (Exception $e) {
-        // error_log('Caught exception: ',  $e->getMessage(), "\n");
-        if (preg_match('/utilized/',$e->getMessage()) ){
-            return "<h4> Awaiting Your Payment </h4>";
+        $existingInvoiceNotes = "";
+    }
+
+    $midtransNotesSeparator = "#PayUrl: ";
+    // Check if Snap URL/token found in invoices notes, then use the existing Snap URL/token
+    if (strpos($existingInvoiceNotes, $midtransNotesSeparator)) {
+        $redirect_url = substr($existingInvoiceNotes, strpos($existingInvoiceNotes, 'midtrans.com'));
+        $extractToken = explode('/', $existingInvoiceNotes);
+        $snapToken = array_pop($extractToken);
+    } else {
+        // Get snap token
+        try {
+            // debug
+            // echo "<pre>";
+            // var_dump($params);
+            // echo "</pre>";
+            // exit();
+            $snap_transaction = Veritrans_Snap::createTransaction($params);
+            $snapToken = $snap_transaction->token;
+            $redirect_url = $snap_transaction->redirect_url;
+            // error_log(" ############# TOKEN ::: ".$snapToken);
+
+            // save snap redirect url to invoice notes
+            try {
+                $whmcsApiCommand = 'UpdateInvoice';
+                $whmcsApiPostParams = array(
+                    'invoiceid' => $invoiceId,
+                    'notes' => $existingInvoiceNotes . $midtransNotesSeparator . $redirect_url
+                );
+                localAPI($whmcsApiCommand, $whmcsApiPostParams);
+            } catch (Exception $e) {
+                echo "error: unable to update invoice notes";
+            }
+
+        } catch (Exception $e) {
+            // error_log('Caught exception: ',  $e->getMessage(), "\n");
+            if (preg_match('/utilized|digunakan/', $e->getMessage())) {
+                // Display payment instruction, prevent payment button from being shown
+                return "Invoice ID has been created on Midtrans previously. Please try to check your email for the previous payment instruction details or open link #PayUrl: on Invoice Notes.";
+            }
         }
```

Comment:
looks like we forgot to handle for other exception, let's add `else` block after existing `if`, maybe just throw the exception so that the error will be visible from merchant. Or call some logging, so the error is logged for troubleshooting.